image/unpack.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package image
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/opencontainers/image-spec/schema"
+	"github.com/pkg/errors"
+)
+
+var (
+	errEOW = fmt.Errorf("end of walk") // error to signal stop walking
+)
+
+// UnpackLayout walks through the file tree given given by src and
+// using the layers specified in the given manifest
+// and unpacks all layers in the given destination directory
+// or returns an error if the unpacking failed.
+func UnpackLayout(src, dest, manifest string) error {
+	return unpack(newPathWalker(src), dest, manifest)
+}
+
+// Unpack walks through the given .tar file and
+// using the layers specified in the given manifest
+// and unpacks all layers in the given destination directory
+// or returns an error if the unpacking failed.
+func Unpack(tarFile, dest, manifest string) error {
+	f, err := os.Open(tarFile)
+	if err != nil {
+		return errors.Wrap(err, "unable to open file")
+	}
+	defer f.Close()
+
+	return unpack(newTarWalker(f), dest, manifest)
+}
+
+func unpack(w walker, dest, manifest string) error {
+	digests, err := diffIDs(w, manifest)
+	if err != nil {
+		return errors.Wrap(err, "error reading diff IDs")
+	}
+
+	return extractLayers(w, digests, dest)
+}
+
+func diffIDs(w walker, manifest string) ([]string, error) {
+	manifest = filepath.Join("manifests", manifest)
```

Comment:
can you change this code to follow the spec here: https://github.com/opencontainers/image-spec/pull/94